### ISSUE TITLE
Add List.chooseV, Seq.tryPickV, etc. for ValueOption

### ISSUE THREAD

**Issue by cmeeren** [fsharp/fslang-suggestions#739]
# Add List.chooseV, Seq.tryPickV, etc. for ValueOption

I propose we add `ValueOption` equivalents of the following functions:

* `List.choose` / `Seq.choose` / `Array.choose`
* `List.pick` / `Seq.pick` / `Array.pick`
* `List.tryPick` / `Seq.tryPick` / `Array.tryPick`
* `List.unfold` / `Seq.unfold` / `Array.unfold`

(For the sake of completeness: Additional functions in these modules that could be implemented too are all the `tryX` functions that only *return* an option, but in this suggestion, I choose to concentrate only on the above since they allocate N objects instead of 1).

The existing way of approaching this problem in F# is... none, really, apart from writing low-level implementations similar to what would go into FSharp.Core.

## Pros and Cons 

The advantages of making this adjustment to F# are:
* Increased performance, since we can use `choose` and `pick` semantics without N allocations
* Better feature parity with `option`

The disadvantages of making this adjustment to F# are:
* Added function bloat (I'd say that for these N-allocating functions it's worth it)

## Extra information

Estimated cost (XS, S, M, L, XL, XXL): S

Related suggestions: None, but see [previous issue in dotnet/fsharp](https://github.com/dotnet/fsharp/issues/6467).

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [x] I or my company would be willing to help implement and/or test this

**Comment by abelbraaksma** [comment]
Can this one be `approved-in-principle` @dsyme? An implementation (https://github.com/dotnet/fsharp/pull/6781) has been ready for some time (since July last year), and I believe this would be nice to have in F# 5.0 if possible :).

I've no problem making an RFC for this to help get this in & documented.

**Comment by gusty** [comment]
I was wondering, for the `try...` functions, isn't the current expectation to return a ref option?

I guess if we stick to that, we should have:

- `tryPick`
- `tryPickV`
- `tryVPick`
- `tryVPickV`

I'm under the impression this claims for a solution at compiler level, like a mechanism to auto-convert/overload (with type inference defaults) between ref and struct options / tuples.

**Comment by abelbraaksma** [comment]
@gusty, I believe the idea is that the V versions use value options all through, and without the V it's the classical way of reference options.

But I agree that some type directed way would be a better solution in the long run, which would allow either type of option in any position. But that poses the problem of how to write different paths if they get auto converted.

**Comment by En3Tho** [comment]
@dsyme @KevinRansom Is there any progress on this? This (with @abelbraaksma 's fixes) would actually be useful in compiler internals. There are places where List.chooce could be easily replaced with valueChoose.

**Comment by travis-leith** [comment]
`chooseV` would be really useful for a numerical experiment I am working on.

**Comment by dsyme** [comment]
I'll mark this as approved in principle.  However I'm aware it's a big addition and should perhaps be done outside FSharp.Core.

@vzarytovskii This may motivate shipping an additional FSharp.Collections.Extensions or something

**Comment by vzarytovskii** [comment]
> I'll mark this as approved in principle. However I'm aware it's a big addition and should perhaps be done outside FSharp.Core.
> 
> @vzarytovskii This may motivate shipping an additional FSharp.Collections.Extensions or something

Yeah, @kevinransom is exploring the possibility for us to ship FSharp.Core as multiple separate packages, this might be a good candidate to start experimenting with it (as well as the same treatment we have for F# core - implicit import, etc).

**Comment by voronoipotato** [comment]
@vzarytovskii am I gonna have to import a second thing for everything to get all the features. If so I'd prefer things to stay as they are unless it makes development easier for you.

**Comment by vzarytovskii** [comment]
> @vzarytovskii am I gonna have to import a second thing for everything to get all the features. If so I'd prefer things to stay as they are unless it makes development easier for you.

No, it will be implicit for users as it is now, just will allow us structure things better, ship things more selectively, and will make it possible to shadow implementations with specific ones for target runtime.

**Comment by vzarytovskii** [comment]
Example: if project is targeting `ns2.0` and implicitly using `FSharp.Core`-`netstandard2.0`, it will be same as now. If project will be targeting `net10.0`, it will load base FSharp.Core (probably `netstandard2.0`/`netstandard2.1`) and one more dll, which will be (for example) FSharp.Core.net10, which may use features from newer BCL.

**Comment by voronoipotato** [comment]
Ah wonderful, this sounds great then :). I recently ran into this actually where it would have been particularly nice to have a valueoption unfold. Part of why I commented on it :P.

**Comment by AntyaDev** [comment]
Hi all,
Any updates on this?

**Comment by T-Gro** [comment]
We can add it for next version of F#.

I am looking for volunteers on reviewing the completeness of the APIs to be added, and perhaps also the wording in doc comments advising users when to use which (since both variants will start appearing for the covered functions).

**Comment by brianrourkeboll** [comment]
Has anyone done a comparison benchmark for these on a recent version of .NET? Last time I checked, the JIT can often stack-allocate short-lived `'T option` nowadays.

**Comment by kerams** [comment]
```fsharp
module ArrayTest =
    let chooseOption (source: _[]) (chooser: _ -> option<_>) =
        let res = ResizeArray<_>()
        for i = 0 to source.Length - 1 do
            match chooser source.[i] with
            | Some v -> res.Add v
            | None -> ()
        res.ToArray()

    let inline chooseOptionInline (source: _[]) ([<InlineIfLambda>] chooser) =
        let res = ResizeArray<_>()
        for i = 0 to source.Length - 1 do
            match chooser source.[i] with
            | Some v -> res.Add v
            | None -> ()
        res.ToArray()

    let chooseValueOption (source: _[]) (chooser: _ -> ValueOption<_>) =
        let res = ResizeArray<_>()
        for i = 0 to source.Length - 1 do
            match chooser source.[i] with
            | ValueSome v -> res.Add v
            | ValueNone -> ()
        res.ToArray()

    let inline chooseValueOptionInline (source: _[]) ([<InlineIfLambda>] chooser) =
        let res = ResizeArray<_>()
        for i = 0 to source.Length - 1 do
            match chooser source.[i] with
            | ValueSome v -> res.Add v
            | ValueNone -> ()
        res.ToArray()

[<MemoryDiagnoser>]
type ArrayChooseBenchmarks() =
    let data = [| 1 .. 10000 |]

    // 1. OPTION baseline
    [<Benchmark(Baseline = true)>]
    member _.Optionchoose () =
        ArrayTest.chooseOption data (fun x ->
            if x % 2 = 0 then Some x else None)

    // 2. OPTION inline lambda
    [<Benchmark>]
    member _.Optionchooseinline () =
        ArrayTest.chooseOptionInline data (fun x ->
            if x % 2 = 0 then Some x else None)

    // 3. VALUEOPTION baseline
    [<Benchmark>]
    member _.ValueOptionchoose () =
        ArrayTest.chooseValueOption data (fun x ->
            if x % 2 = 0 then ValueSome x else ValueNone)

    // 4. VALUEOPTION inline lambda
    [<Benchmark>]
    member _.ValueOptionchooseinline () =
        ArrayTest.chooseValueOptionInline data (fun x ->
            if x % 2 = 0 then ValueSome x else ValueNone)

BenchmarkDotNet.Running.BenchmarkRunner.Run (
    typeof<ArrayChooseBenchmarks>,
    DefaultConfig.Instance.WithOption(ConfigOptions.JoinSummary, true))
|> ignore
```

```

BenchmarkDotNet v0.16.0-nightly.20260608.560, Windows 11 (10.0.26200.8655/25H2/2025Update/HudsonValley2)
AMD Ryzen 9 7900 3.70GHz, 1 CPU, 24 logical and 12 physical cores
Memory: 31.12 GB Total, 14.49 GB Available
.NET SDK 11.0.100-preview.5.26302.115
  [Host]     : .NET 11.0.0 (11.0.0-preview.5.26302.115, 11.0.26.30315), X64 RyuJIT x86-64-v4 DEBUG
  DefaultJob : .NET 11.0.0 (11.0.0-preview.5.26302.115, 11.0.26.30315), X64 RyuJIT x86-64-v4


```
| Method                  | Mean      | Error     | StdDev    | Ratio | RatioSD | Gen0    | Gen1   | Allocated | Alloc Ratio |
|------------------------ |----------:|----------:|----------:|------:|--------:|--------:|-------:|----------:|------------:|
| Optionchoose            | 13.672 μs | 0.2059 μs | 0.1720 μs |  1.00 |    0.00 | 12.2833 | 1.6327 | 201.04 KB |        1.00 |
| Optionchooseinline      | 13.060 μs | 0.2155 μs | 0.1911 μs |  0.96 |    0.02 | 12.2833 | 1.6327 | 201.04 KB |        1.00 |
| ValueOptionchoose       |  6.162 μs | 0.1178 μs | 0.1310 μs |  0.45 |    0.01 |  5.1270 |      - |  83.85 KB |        0.42 |
| ValueOptionchooseinline |  6.082 μs | 0.0387 μs | 0.0323 μs |  0.44 |    0.01 |  5.1270 |      - |  83.85 KB |        0.42 |

I am quite surprised `chooseOptionInline` didn't use stack-allocated options... wonder why. In any case, this is a win for `chooseV`, where lots of options are created. Don't think we really need `pickV` though (there's at most one `Some` allocation regardless of the collection size).

**Comment by Happypig375** [comment]
How much can stack allocated reference types save us memory and performance here?

**Comment by kerams** [comment]
I'd expect `Optionchooseinline` to allocate the same amount as the voption functions, but `ValueOptionchooseinline` should be ever so slightly faster if struct promotion kicks in. Just an educated guess though:) A lot depends on stack allocation eligibility escape analysis (`inline` and `[<InlineIfLambda>]` in `chooseOptionInline` should at first glance satisfy it, but alas) and other heuristics.

**Comment by bartelink** [comment]
For completeness, it would be good to add variants for all standard functions that yield `option`, e.g. `tryHeadV`

**Comment by bartelink** [comment]
Some semi related (ok, probably tangential) ideas triggered by this (please no big thread here, just vote on this post as indicated if anyone thinks the idea has enough merit or stupidity to warrant an independent issue where it can be debated in full).

~(This is already in FSharp.Core, my bad)~
~A `module ValueOption` with: (👍  reaction and I'll add a suggestion)~
~- probably 1:1 as per Option (I often want `map`)~
~- maybe converters, e.g. `ValueOption.ofOption`, `ValueOption.toOption`, `Option.toV` ?~

----

A `module ValueTuple` with ( :eyes: reaction and I'll expand out to a suggestion) 
- ofTuple
- ofKvp, toKvp
- (maybe) `groupWith`:
   ```fsharp
   let inline groupWith ([<InlineIfLambda>] f) xs =
       Seq.groupBy fst xs
       |> Seq.map (fun (k, xs) -> struct (k, xs |> Seq.map snd |> f))
     ```

----

Aliases:

We have exactly one pair (`where` vs `filter`), and don't want any more (though see `any` below!)

----

Inverses (🚀 and I'll write up a suggestion)

On the other hand, frequently useful inverses make code more legible and expressive.

The obvious cost / reason not to do it is that it means there are more things in the box to learn / understand

 | Proposed inverse | Impl sketch | Notes | 
 |--|--|--| 
 | forall | (not << exists) (not << predicate) | Already in the box |
 | exists | (not << forall) (not << predicate) | Already in the box |
 | isNotEmpty | not << isEmpty | Could also be called `any`, but `Not` makes purpose/redundancy clear and puts it nearby in completion list |
 | reject | filter (not << predicate) | Could perhaps be called `drop`, but that's normally a non-throwing `take` |
 | none | forall (not << predicate) | |

**Comment by xperiandri** [comment]
@dsyme also `fstv` and `sndv` functions are needed.
1. Can the implementation of this suggestion include them?
2. What names would you approve? `fstv` vs `fstV`?

**Comment by kerams** [comment]
<img width="409" height="274" alt="Image" src="https://github.com/user-attachments/assets/2c94247b-69ef-4f66-900e-c3dbeac0bb63" />

I'd love to see something like this (it solves struct vs ref and cardinality issues), but I think that ship might have sailed already, and perhaps the fact that this would break the role of `fst` and `snd` in type inference is undesirable (or maybe with `open FSharp.Core.ModernPrelude`?:)).

**Comment by brianrourkeboll** [comment]
@kerams https://github.com/fsharp/fslang-suggestions/issues/739#issuecomment-4729523002

> I am quite surprised `chooseOptionInline` didn't use stack-allocated options... wonder why.

https://github.com/dotnet/runtime/issues/104936: I guess no form of https://github.com/dotnet/runtime/pull/106526 has made it in, so it doesn't work in loops.

**Comment by xperiandri** [comment]
@kerams maybe add this as a separate suggestion? I like your idea very much!

**Comment by T-Gro** [comment]
> I'd love to see something like this (it solves struct vs ref and cardinality issues)

```[<AutoOpenIfLangVersion("11.0")>]```

But I think even when versioned, the impact on type inference is too large.
We would need a mechanism to tie-break the overload instead of raising ambiguity warnings.

**Comment by kerams** [comment]
`OverloadResolutionPriority` wouldn't work here? I can't build main and try it right now.

**Comment by xperiandri** [comment]
@kerams I can help you. Reach me in X

**Comment by T-Gro** [comment]
@kerams :
I think you are right, this could work:

```fsharp
[<AutoOpen>]
type TupleAccessors =
    [<OverloadResolutionPriority(1)>]
    static member fst ((a, _)) = a
    static member fst (struct (a, _)) = a
    static member fst ((a, _, _)) = a
    static member fst (struct (a, _, _)) = a

    [<OverloadResolutionPriority(1)>]
    static member snd ((_, b)) = b
    static member snd (struct (_, b)) = b
    static member snd ((_, b, _)) = b
    static member snd (struct (_, b, _)) = b

let f x = fst x // defaults to ref-pair
let myst = fst // ref pair
```
